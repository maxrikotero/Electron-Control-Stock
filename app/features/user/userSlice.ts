import { createSlice } from '@reduxjs/toolkit';
// eslint-disable-next-line import/no-cycle
import { AppThunk, RootState } from '../../store';

const userSlice = createSlice({
  name: 'user',
  initialState: { sessionData: {} },
  reducers: {
    set: (state, action) => {
      state.sessionData = action.payload;
    },
  },
});

export const { set } = userSlice.actions;

export default userSlice.reducer;

export const selectCount = (state: RootState) => state.user.sessionData;
