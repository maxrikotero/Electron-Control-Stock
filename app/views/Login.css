* {
  margin: 0;
}

.login__content {
  height: 100%;
  color: white;
  overflow: hidden;
  background-color: #111;
}

.login {
  background-color: white;
  top: 0;
}
.login_header {
  display: flex;
  justify-content: center;
}

.login__logo {
  height: 28vh;
}

.form_content {
  margin-top: 7%;
  height: 100%;
}

.form_body {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.input__content {
  width: 450px;
}
