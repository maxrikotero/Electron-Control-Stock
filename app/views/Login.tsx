import React, {
  useState,
  useRef,
  useEffect,
  Fragment,
  ReactElement,
} from 'react';
import { Redirect } from 'react-router-dom';
import NotificationSystem from 'react-notification-system';
import {
  Grid,
  Row,
  Col,
  FormGroup,
  ControlLabel,
  FormControl,
} from 'react-bootstrap';
import { Formik } from 'formik';
import { useDispatch } from 'react-redux';
import '../assets/css/login.css';
import Button from '../components/CustomButton/CustomButton';
import { style } from '../variables/Variables';
import useApiUrl from '../hooks/useApiUrl';
import { set } from '../features/user/userSlice';
import logo from '../assets/img/logo.jpeg';
import './Login.css';

const Login = () => {
  const dispatch = useDispatch();

  const apiUrl: string = useApiUrl();

  const notificationSystem = useRef<HTMLInputElement>(null);

  const [redirection, setRedirection] = useState<boolean>(false);

  useEffect(() => {}, []);
  const addNotification = ({
    position,
    message,
    color,
  }: {
    position: string;
    message: ReactElement;
    color: number;
  }) => {
    var level;
    switch (color) {
      case 1:
        level = 'success';
        break;
      case 2:
        level = 'warning';
        break;
      case 3:
        level = 'error';
        break;
      case 4:
        level = 'info';
        break;
      default:
        break;
    }
    const notification: any = notificationSystem.current;

    notification.addNotification({
      title: <span data-notify="icon" className="pe-7s-bell" />,
      message: <div>{message}</div>,
      level: level,
      position: position,
      autoDismiss: 15,
    });
  };

  return (
    <Fragment key="loginFragment">
      <NotificationSystem ref={notificationSystem} style={style} />
      {redirection && <Redirect to="/admin/principal" />}
      <div className="login__content">
        <div className="login">
          <div className="login_header">
            <img className="login__logo" src={logo} alt="..." />
          </div>
        </div>
        <div className="form_content">
          <Formik
            initialValues={{ email: '', password: '' }}
            validate={(values) => {
              const errors: any = {};
              if (!values.password) {
                errors.password = 'Requerido';
              }
              if (!values.email) {
                errors.email = 'Requerido';
              } else if (
                !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)
              ) {
                errors.email = 'Email Invalido';
              }
              return errors;
            }}
            onSubmit={(values, { setSubmitting }) => {
              fetch(`${apiUrl}/users/signin`, {
                method: 'POST',
                body: JSON.stringify(values),
                headers: {
                  Accept: 'application/json',
                  'Content-Type': 'application/json',
                },
              })
                .then((res) => res.json())
                .then(({ success, message, token, userSession }) => {
                  setSubmitting(false);
                  if (success) {
                    dispatch(set(userSession));
                    localStorage.setItem('token', token);
                    setRedirection((prev) => !prev);
                  } else {
                    addNotification({
                      position: 'tr',
                      message: <div>{message}</div>,
                      color: 3,
                    });
                  }
                })
                .catch(() => {
                  setSubmitting(false);

                  addNotification({
                    position: 'tr',
                    message: <div>Login Error</div>,
                    color: 3,
                  });
                });
            }}
          >
            {({
              values,
              errors,
              touched,
              handleChange,
              handleSubmit,
              isSubmitting,
            }) => (
              <form className="form_body" onSubmit={handleSubmit}>
                <div className="input__content">
                  <FormGroup controlId="emailControl">
                    <ControlLabel
                      style={{ fontWeight: 'initial', fontSize: '15px' }}
                    >
                      Email
                    </ControlLabel>
                    <FormControl
                      type="email"
                      name="email"
                      onChange={handleChange}
                      bsClass="form-control"
                      value={values.email}
                    />
                  </FormGroup>
                  <span style={{ color: 'red' }}> {errors.email}</span>
                </div>

                <div className="input__content">
                  <FormGroup controlId="passControl">
                    <ControlLabel
                      style={{ fontWeight: 'initial', fontSize: '15px' }}
                    >
                      Password
                    </ControlLabel>
                    <FormControl
                      type="password"
                      name="password"
                      onChange={handleChange}
                      bsClass="form-control"
                      value={values.password}
                    />
                  </FormGroup>
                  <span style={{ color: 'red' }}>
                    {' '}
                    {errors.password && touched.password && errors.password}
                  </span>
                </div>

                <div style={{ textAlign: 'center', marginTop: '30px' }}>
                  <Button
                    bsStyle="info"
                    fill
                    type="submit"
                    disabled={isSubmitting}
                  >
                    Iniciar Sesi√≥n
                  </Button>
                </div>

                <div className="clearfix" />
              </form>
            )}
          </Formik>
        </div>
      </div>
    </Fragment>
  );
};

export default Login;
