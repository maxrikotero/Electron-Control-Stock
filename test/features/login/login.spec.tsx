/* eslint react/jsx-props-no-spreading: off, @typescript-eslint/ban-ts-comment: off */
import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { waitFor } from '@testing-library/react';
import Login from '../../../app/views/Login';
import * as userSlice from '../../../app/features/user/userSlice';

Enzyme.configure({ adapter: new Adapter() });
// jest.useFakeTimers();

interface useData {
  name: string;
}

function setup(
  preloadedState: { user: { sessionData: useData } } = {
    user: { sessionData: { name: 'Maxi' } },
  }
) {
  const store = configureStore({
    reducer: { user: userSlice.default },
    preloadedState,
  });

  const shallowComponent = shallow(
    <Provider {...{ store }}>
      <Router>
        <Login />
      </Router>
    </Provider>
  );

  const getWrapper = () =>
    mount(
      <Provider {...{ store }}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

  const component = getWrapper();

  return {
    store,
    component,
    shallowComponent,
  };
}

describe('Login Component', () => {
  it("should render correctly in 'debuge' mode", () => {
    const { store } = setup();

    const component = shallow(
      <Provider {...{ store }}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    expect(component).toMatchSnapshot();
  });
  it('should update email field on change', async () => {
    const { component } = setup();

    const emailInput = component.find("input[name='email']");

    emailInput.simulate('change', {
      persist: () => {},
      target: {
        name: 'email',
        value: 'email@test.com',
      },
    });

    await waitFor(() => {
      expect(emailInput.html()).toMatch('email@test.com');
    });
  });

  it('should update password field onChange', async () => {
    const { component } = setup();

    const passwordInput = component.find("input[name='password']");

    passwordInput.simulate('change', {
      persist: () => {},
      target: {
        name: 'password',
        value: '1234',
      },
    });

    await waitFor(() => {
      expect(passwordInput.html()).toMatch('1234');
    });
  });

  // it('should return error form invalid email address', async () => {
  //   const { store } = setup();

  //   const shallowComponent = shallow(
  //     <Provider {...{ store }}>
  //       <Router>
  //         <Login />
  //       </Router>
  //     </Provider>,
  //     {
  //       wrappingComponent: Provider,
  //       wrappingComponentProps: { store },
  //     }
  //   );

  //   const loginForm = (props: {
  //     errors: {};
  //     touched: { email: boolean };
  //     isSubmitting: boolean;
  //   }) =>
  //     shallowComponent.find(Login).dive().find(Formik).renderProp('children')(
  //       props
  //     );

  //   const formWithInvalidEmailErrors = loginForm({
  //     errors: {
  //       email: 'Invalid Email Address',
  //     },
  //     touched: { email: true },
  //     isSubmitting: false,
  //   });

  //   await waitFor(() => {
  //     expect(formWithInvalidEmailErrors.html()).toMatch(
  //       /Invalid Email Address/
  //     );
  //   });
  // });
});
