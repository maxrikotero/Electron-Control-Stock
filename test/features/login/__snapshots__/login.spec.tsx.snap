// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should render correctly in 'debuge' mode 1`] = `ShallowWrapper {}`;
